src/VizzyUtils/QueryIngestion/VizzyPrepare.ts

```diff
@@ -0,0 +1,125 @@
+   */
+  validateInputShape: (data: any, addError: any, queryLength: number, strictLength: boolean) => void,
+  /**
+   * Ensure that the input data is of a shape which can be properly prepared
+   * @param data - data obj from Looker
+   * @param queryResponse - queryResponse obj from Looker
+   * @param rules - some options for error handling/boundary checking
+   * @param addError - function for reporting an error back to Looker
+   */
+  validatePrepareShape: (data: any, queryResponse: VisQueryResponse, rules: QueryFormatterRules, addError: any) => void,
+}
+
+export const VizzyQueryFormatter: QueryFormatter = {
+  validateInputShape(data, addError, queryLength, strictLength) {
+    if (strictLength) {
+      if (data.length !== queryLength) {
+        addError({title: "Invalid input", message: `Expected ${data.length} ${data[0][0].field_category}s, found ${queryLength}.`})
+      }
+    }
+  },
+  validatePrepareShape(data, queryResponse, rules, addError) {
+    const actualDims = queryResponse.fields.dimension_like.length
+    const actualMeas = queryResponse.fields.measure_like.length
+    if (data.length === 0) {
+      addError({title: "No data"})
+    }
+    if (rules.strictDims && actualDims !== rules.numDims) {
+      addError({title: "Invalid input", message: `Expected ${rules.numDims} dimensions, found ${actualDims}.`})
+    }
+    else if (rules.strictDims && actualMeas !== rules.numMeas) {
+      addError({title: "Invalid input", message: `Expected ${rules.numMeas} measures, found ${actualMeas}.`})
+    }
+  },
+  getQueryMetadata(queryFields, columns) {
+    const filtered = queryFields.filter((d: any, i: number) => {
+      return i <= columns - 1
+    })
+    return filtered.map((d: any) => {
+      // this is where queryResponse metadata can be brought over to each Cell
+      return {field_name: d.name, field_category: d.category, label: d.label_short}
+    })
+  },
+  prepare(data, config, queryResponse, vis, rules) {
+    this.validatePrepareShape(data, queryResponse, rules, vis.addError)
+    
+    const dims = this.getQueryMetadata(queryResponse.fields.dimension_like, rules.numDims || 0)
+    const dimsData = dims.map((dim: any) => {
+      return data.map((row: any) => {
+        return {
+          ...row[dim.field_name],
+          ...dim
+        }
+      })
+    })
+    this.validateInputShape(dimsData, vis.addError, queryResponse.fields.dimension_like.length, rules.strictDims || false)
+    
+    const meas = this.getQueryMetadata(queryResponse.fields.measure_like, rules.numMeas || 0)
+    const measData = meas.map((mea: any) => {
+      return data.map((row: any) => {
+        return {
+          ...row[mea.field_name],
+          ...mea
+        }
+      })
+    })
+    this.validateInputShape(measData, vis.addError, queryResponse.fields.measure_like.length, rules.strictMeas || false)
+    return {
+      dimensions: [...dimsData],
+      measures: [...measData],
+      ...rules
+    }
+  },
+}
```

Comment:
Instead of a series of utils in an object, this shape might be better served by a class. It wouldn't take too much, since `prepare` is essentially your constructor.
You could then cache the various values on the class instance rather than just returning data the one time.

What you're doing here is very similar to what we did way back in QueryResultTable
https://github.com/looker/helltool/blob/master/lib/helltool/assets/visualizations/query_result_table.js

```
// visData is now a class instance with methods and data
const vizData = new Vizzy(data, config, queryResponse, vis, {...config})
```